File: damon/src/main/java/com/zyhang/damon/support/MvpAppCompatActivity.java
package com.zyhang.damon.support;

import android.os.Bundle;

import com.zyhang.damon.MvpPresenter;
import com.zyhang.damon.factory.PresenterGetter;
import com.zyhang.damon.PresenterLifecycleDelegate;
import com.zyhang.damon.factory.ReflectionPresenterFactory;

import java.util.List;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

/**
 * ProjectName:Damon
 * Description:
 * Created by zyhang on 2017/4/28.23:24
 * Modify by:
 * Modify time:
 * Modify remark:
 */

public class MvpAppCompatActivity<P extends MvpPresenter> extends AppCompatActivity implements PresenterGetter<P> {

    private static final String PRESENTER_STATE_KEY = "presenter_state";
    private PresenterLifecycleDelegate<P> mPresenterDelegate =
            new PresenterLifecycleDelegate<>(ReflectionPresenterFactory.fromViewClass(this, getClass()));

    @Nullable
    @Override
    public List<? extends MvpPresenter> getPresenters() {
        return mPresenterDelegate.getPresenters();
    }

    @Nullable
    @Override
    public P getPresenter() {
        return mPresenterDelegate.getPresenter();
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mPresenterDelegate.dispatchCreate(this, getIntent().getExtras(), savedInstanceState, null != savedInstanceState ? savedInstanceState.getBundle(PRESENTER_STATE_KEY) : null);
        mPresenterDelegate.dispatchCreateView();
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putBundle(PRESENTER_STATE_KEY, mPresenterDelegate.getPresenterSaveState());
    }

    @Override
    public void onStart() {
        super.onStart();
        mPresenterDelegate.dispatchStart();
    }

    @Override
    public void onResume() {
        super.onResume();
        mPresenterDelegate.dispatchResume();
    }

    @Override
    public void onPause() {
        super.onPause();
        mPresenterDelegate.dispatchPause();
    }

    @Override
    public void onStop() {
        super.onStop();
        mPresenterDelegate.dispatchStop();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        mPresenterDelegate.dispatchDestroyView();
        mPresenterDelegate.dispatchDestroy(!isChangingConfigurations());
    }
}

File: damon-kotlin/src/main/kotlin/com/zyhang/damon/kotlin/support/MvpAppCompatActivity.kt
package com.zyhang.damon.kotlin.support

import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import com.zyhang.damon.kotlin.*

/**
 * ProjectName:Damon
 * Description:
 * Created by zyhang on 2017/4/28.下午5:55
 * Modify by:
 * Modify time:
 * Modify remark:
 */

open class MvpAppCompatActivity<out P : MvpPresenter<MvpView>> : AppCompatActivity(), ViewWithPresenter<P>, MvpView {

    private val PRESENTER_STATE_KEY = "presenter_state"
    val mPresenterDelegate: PresenterLifecycleDelegate<P> =
            PresenterLifecycleDelegate(ReflectionPresenterFactory.Companion.fromViewClass(javaClass.kotlin))

    override fun getPresenter(): P? {
        return mPresenterDelegate.presenter()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mPresenterDelegate.OnCreate(this, intent.extras, savedInstanceState?.getBundle(PRESENTER_STATE_KEY))
    }

    override fun onStart() {
        super.onStart()
        mPresenterDelegate.onStart()
    }

    override fun onSaveInstanceState(outState: Bundle?) {
        super.onSaveInstanceState(outState)
        outState?.putBundle(PRESENTER_STATE_KEY, mPresenterDelegate.onSaveInstanceState())
    }

    override fun onResume() {
        super.onResume()
        mPresenterDelegate.onResume()
    }

    override fun onPause() {
        mPresenterDelegate.onPause()
        super.onPause()
    }

    override fun onStop() {
        mPresenterDelegate.onStop()
        super.onStop()
    }

    override fun onDestroy() {
        mPresenterDelegate.onDestroy(!isChangingConfigurations)
        super.onDestroy()
    }
}